### PR TITLE
Smartsurvey api

### PR DESCRIPTION
**Pull request by shilohd**
Please review the injest script, which retrieves the data from SmartSurvey using the API.

### CODE REVIEW

**Comment by alexander-newton** [comment]
Thanks for this; can confirm that I can reproduce.

- It would be great if the ingest also checked the hash and failed on not being able to reproduce the hash. You might want to actually download the CSV to a temporary folder first; check the hash and then read it in.
- Please can you write some documentation about how to get the smartsurvey API token and secret for other users and how to save it to environment variables in R
- Please can you turn this into a function in `R/ingest.R` - the package folder, make it installable and call it at the start of the `data_cleaning.R` script when Sophie is comfortable with that.

**Comment by shilohd** [comment]
Packages ingest functions. Please review.

File: carsurvey2.Rcheck/00_pkg_src/carsurvey2/R/tidy_ingest_data.R [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  
+  colnames(data) <- new_colnames
+  
+  # Drop empty rows
+  data <- data[!is.na(data$userID), ] 
+
+  # Correct strings
+  data <- data.frame(
+    lapply(data, function(x) {
+      gsub("Don@SQ@t Know", "Don't Know", x)
```
Comment (foster999) [inline carsurvey2.Rcheck/00_pkg_src/carsurvey2/R/tidy_ingest_data.R:79]: If this encoding error occurs in other places, we might use a more general patter?
```suggestion
      gsub("@SQ@", "'", x)
```
No worries if this is the only case though

File: carsurvey2.Rcheck/00_pkg_src/carsurvey2/R/ingest.R [hunk 2]
```diff
@@ -0,0 +1,72 @@
+#'@title Download smartsurvey export via the API
```
Comment (foster999) [inline carsurvey2.Rcheck/00_pkg_src/carsurvey2/R/ingest.R:1]: Huge fan of this function, implemented really well! 👍

File: carsurvey2.Rcheck/00_pkg_src/carsurvey2/R/tidy_ingest_data.R [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    factor(q_freqs$q_numbers, levels = unique(q_numbers))
+  q_freqs <- q_freqs[order(q_freqs$q_numbers),]
+  
+  new_colnames = Map(
+    function(name, freq) {
+      if (freq == 1) {
+        return(as.character(name))
+      } else {
+        return(c(paste0(rep(name, freq),
+                        c(
```
Comment (foster999) [inline carsurvey2.Rcheck/00_pkg_src/carsurvey2/R/tidy_ingest_data.R:58]: Please add a comment to describe simply what decisions are made in this mapping.

File: carsurvey2.Rcheck/00_pkg_src/carsurvey2/R/ingest.R [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+#'
+#'@description Download the exported data from smartsurvey using the API and convert binary csv file to a data frame.
+#'
+#'@param survey the survey ID (character string/numeric). Defaults to "790800".
+#'@param export the export ID (character string/numeric). Defaults to "1438876".
+#'@param token the API token (character string). Loaded from environment variable by default.
+#'@param secret the secret API token (character string). Loaded from environment variable by default.
+#'@param check_hash a hash to check the downloaded data aginst or FALSE to skip the check (character string/logical). Defaults to final export hash.
+#'
```
Comment (foster999) [inline carsurvey2.Rcheck/00_pkg_src/carsurvey2/R/ingest.R:10]: Reference the related `convert_raw` function, as this should be used to process the output.

File: carsurvey2.Rcheck/00_pkg_src/carsurvey2/R/tidy_ingest_data.R [hunk 5]
```diff
@@ -0,0 +1,84 @@
+#'@title Tidy the ingested data from the smartsurvey API
+#'
+#'@description Tidy column names and remove empty rows from the ingested data. 
+#' The question numbers returned are only nested to one level, e.g. Q1.1, Q1.2, etc.
+#'
+#'@param data the data returned by the ingest function (data.frame)
+#'
+#'@return the tidied data (data.frame)
+#'
```
Comment (foster999) [inline carsurvey2.Rcheck/00_pkg_src/carsurvey2/R/tidy_ingest_data.R:9]: Reference `convert_raw`, as this will always be run on the output of that function.

File: carsurvey2.Rcheck/00_pkg_src/carsurvey2/R/tidy_ingest_data.R [hunk 6]
```diff
@@ -0,0 +1,84 @@
+#'@title Tidy the ingested data from the smartsurvey API
```
Comment (foster999) [inline carsurvey2.Rcheck/00_pkg_src/carsurvey2/R/tidy_ingest_data.R:1]: Would this work with any number of nested levels? If not, specify what levels it will work for.

File: carsurvey2/tests/testthat/test-tidy_ingest_data.R [hunk 7]
```diff
@@ -0,0 +1,55 @@
+# No nested survey questions
```
Comment (foster999) [inline carsurvey2/tests/testthat/test-tidy_ingest_data.R:1]: Would it also be worth testing the dropping of empty rows and numbering of duplicated columns? If these are important features of the function.

**Review by foster999 — APPROVED** [review]
# Review

Looks great overall. Have made some minor suggestions, but would merge it once you have applied any of these which you agree with 👍 

## Package

- [ ] Add the `.Rcheck` directory to `.gitignore. These files inflate the repo and make it difficult to review. E.g. there are 56 files in this PR.
- [ ] Add the`tar.gz` to gitignore. These are either built manually by users or uploaded/downloaded to a package release repository (e.g. CRAN or GitHub releases).

## Code Review

* Great input validation and error handling.
* Tests cover all realistic use cases.
* Please action other points on actual R files, rather than .Rcheck examples that I've accidentally  commented on 😄 

## General Site

- [ ]  It would be good to record the site/`build_site.R` dependencies in some way (e.g. `plotly` is not installed with the package).

- [ ] `_site.yml` should end with a blank newline. `build_sire.R` currently warns about incomplete lines when reading it.

File: carsurvey2/tests/testthat/test-tidy_ingest_data.R (hunk 7 above)
Comment (shilohd) [reply carsurvey2/tests/testthat/test-tidy_ingest_data.R:1]: Dropping of empty rows is tested by the checking the number of rows in the output, i.e. if more or fewer than 2 rows are returned that means the function failed to remove the empty rows in the dataset. The numbering of duplicate columns is only needed to ensure multiple columns don't have the same name when the answers to a question are spread over multiple column, but seems to work well from the tests that are already there.

File: carsurvey2.Rcheck/00_pkg_src/carsurvey2/R/ingest.R (hunk 2 above)
Comment (shilohd) [reply carsurvey2.Rcheck/00_pkg_src/carsurvey2/R/ingest.R:1]: Thanks

File: carsurvey2.Rcheck/00_pkg_src/carsurvey2/R/tidy_ingest_data.R (hunk 1 above)
Comment (shilohd) [reply carsurvey2.Rcheck/00_pkg_src/carsurvey2/R/tidy_ingest_data.R:79]: Apostrophes don't appear to come through the API this way in any other instances.
